File: src/chanhyun/week3/보석 쇼핑.py
from collections import defaultdict

def solution(gems):
    answer = []
    length = len(gems)+1
    gem_num = len(set(gems))
    basket = defaultdict(int)
    l, r = 0, 0
    while r < len(gems):
        basket[gems[r]] += 1
        r += 1
        
        if len(basket) >= gem_num:
            while l < r:
                if basket[gems[l]] > 1:
                    basket[gems[l]] -= 1
                    l += 1
                else:
                    break
            if length > r-l:
                length = r-l
                answer = [l+1, r]
    return answer

File: src/taeeun/week3/보석 쇼핑.js
function solution(gems) {
    var answer = [];
    let gemMap=new Map();
    const gemSize=new Set(gems).size;
    gems.forEach((gem, index)=>{
        gemMap.delete(gem)
        gemMap.set(gem, index+1);
        if (gemMap.size===gemSize) {
            const minArray=[gemMap.values().next().value, index+1];
            if (answer.length===0 || answer[1]-answer[0]>minArray[1]-minArray[0]){
                answer=minArray;
            }
        }
    })
    return answer;
}
